crypto3/libs/blueprint/include/nil/blueprint/bbf/circuit_builder.hpp

```diff
@@ -0,0 +1,668 @@
+                            if (lookup_table_names.find(table_name) == lookup_table_names.end()) {
+                                if (gates.dynamic_lookup_tables.find(table_name) != gates.dynamic_lookup_tables.end()) {
+                                    bp.reserve_dynamic_table(table_name);
+                                } else {
+                                    bp.reserve_table(table_name);
+                                }
+                                lookup_table_names.insert(table_name);
+                            }
+                            std::size_t table_index = bp.get_reserved_indices().at(table_name);
+                            lookup_gate.push_back({table_index, single_lookup_constraint.second});
+                        }
+
+                        bp.add_lookup_gate(selector_id, lookup_gate);
+                    }
+
+                    // compatibility layer: dynamic lookup tables - continued
+                    for(const auto& [name, area] : gates.dynamic_lookup_tables) {
+                        bp.register_dynamic_table(name);
+
+                        std::size_t selector_index = area.second;
+
+                        //crypto3::zk::snark::plonk_lookup_table<FieldType> table_specs;
+                        plonk_lookup_table table_specs;
+                        table_specs.tag_index = selector_index;
+                        table_specs.columns_number = area.first.size();
+                        std::vector<var> dynamic_lookup_cols;
+                        for(const auto& c : area.first) {
+                            // TODO: does this make sense?!
+                            dynamic_lookup_cols.push_back(var(c, 0, false, var::column_type::witness));
+                        }
+                        table_specs.lookup_options = {dynamic_lookup_cols};
+                        bp.define_dynamic_table(name,table_specs);
+                    }
+
+                    // this is where we pack lookup tables into constant columns and assign them selectors
+                    std::size_t usable_rows = std::pow(2, std::ceil(std::log2(rows_amount))) - 1; // assure minimum inflation after padding
+                    const auto &lookup_table_ids = bp.get_reserved_indices();
+                    const auto &lookup_tables = bp.get_reserved_tables();
+                    const auto &dynamic_tables = bp.get_reserved_dynamic_tables();
+
+                    std::vector<std::string> ordered_table_names =
+                        nil::crypto3::zk::snark::get_tables_ordered_by_rows_number<FieldType>(lookup_tables);
+
+                    std::size_t start_row = 1;
+                    std::vector<plonk_lookup_table> bp_lookup_tables(lookup_table_ids.size());
+                    std::size_t start_constant_column = presets.constants_amount();
+                    std::size_t prev_columns_number = 0;
+                    int full_selector_id = 0;
```

Comment:
Are we sure 0 here is good enough a merker? Maybe change to -1? AFAIK 0 is a proper selector id.

Reply:
Fixed. I had the same idea (wherefrom int is the type) but then reverted for some reason.